### ISSUE TITLE
Example of using key_padding_mask for flash attention v2

### ISSUE THREAD

**Issue by YJYJLee** [Dao-AILab/flash-attention#530]
1. Could anyone provide a good example of using key_padding_mask for flash attention v2?
key_padding_mask is used to mask out the keys (shape [B, 1, 1, N])

2. Is it possible to use key_padding_mask and causal mask together?

**Comment by sz128** [comment]
https://github.com/lm-sys/FastChat/blob/main/fastchat/train/llama_flash_attn_monkey_patch.py

**Comment by franp9am** [comment]
Thank you @sz128 Based on the link, here are a few lines of my adaptation (I didn't use causal mask, only key_padding_mask) in my MHA forward...

# self attention MHA
 def forward(
        self,
        x,
        key_padding_mask=None
    ):
        assert x.ndim == 3

        b, l, d = x.shape
        
        qkv = self.input_proj(x)  # shape (batch, L, 3 x embed_dim)

        # my convention as in torch: True for padded, False for attended
        if key_padding_mask is None:
            key_padding_mask = torch.ones(b, l).bool().to(q.device)
        else:
            key_padding_mask = ~key_padding_mask

        qkv, indices, cu_q_lens, max_s, _ = unpad_input(qkv, key_padding_mask)        
        qkv = qkv.view(-1, 3, self.num_heads, self.head_dim)

        output_unpad = flash_attn_varlen_qkvpacked_func(
                qkv, cu_q_lens, max_s, self.dropout, softmax_scale=None, causal=False
        )        
        output_unpad = output_unpad.reshape(-1, self.embed_dim)        
        output = pad_input(output_unpad, indices, b, l)
        return self.out_proj(output)

**Comment by Lau-Jonathan** [comment]
@franp9am Thank you for your explanation, but I would like to know if you have debugged it in your own environment to ensure that this code solves the issue with supporting key_padding_mask in FlashAttention.

**Comment by franp9am** [comment]
@Lau-Jonathan  Yes, I did -- was comparing it with torch MHA an checking it gives the same result. But you should double-check too, no guarantees :-) 

Also, this pad_unpad and flash_attn_varlen_qkvpacked_func are heavily dependent on flash-attn version. I would be of course much cleaner if it was natively integrated to main functions api. There is no fundamental reason for not having it, and seems relatively simple -- I guess they will do it sooner or later.

**Comment by Lau-Jonathan** [comment]
@franp9am Thank you very much for your reply. I would like to know if the code above still works when the entire window consists of invalid tokens, i.e., when cu_q_lens=0. I’m not sure if you’ve encountered a similar issue. This is just an additional question, but the main issue has already been addressed. Thanks again!

**Comment by Lau-Jonathan** [comment]
> [@franp9am](https://github.com/franp9am) Thank you very much for your reply. I would like to know if the code above still works when the entire window consists of invalid tokens, i.e., when cu_q_lens=0. I’m not sure if you’ve encountered a similar issue. This is just an additional question, but the main issue has already been addressed. Thanks again!

I reviewed the source code, and when the valid length is 0, it should throw an error because the extracted QKV will be an empty tensor, which cannot be used for computation.
